### PR TITLE
VersionSpec: expose full regex power

### LINKED ISSUE #2716: Version pinning misses capability for Regular Expressions
I would like to restrict a package's version to release versions through pinning, excluding pre-release builds.
I think the only way to achieve this is using a regex for the version specification in the `pinned` file. But since there is no way to explicitly force a line in the pinned file to be marked a regex, [we can only get a regex-matching `VersionSpec` by satisfying this line in it's constructor](https://github.com/conda/conda/blob/231e9db898b3d7720b49e9e2050a88be6978fd38/conda/version.py#L319):
```
    elif '*' in spec.rstrip('*'):
```
So we would have to put multiple stars in the version spec but make sure that the resulting regex is still what we want to match.. plus, there's all kind of implicit replacementsdone to our regex, so this is not feasible I think.

Consider this example:
```
((maintenance_1.0.x)) wintermute:~$ conda search basemap
Fetching package metadata ...........
basemap                      1.0.6                np17py27_0  defaults        
                          .  1.0.7               np110py27_0  defaults        
                             1.0.8.dev           np110py27_0  conda-forge/linux-64 
                          *  1.0.8.dev0          np110py27_1  conda-forge/linux-64 
```
Let's assume I want to pin to any release version build of `1.0` series, excluding the "*.dev*" pre-release packages. I don't see a way to do that..

How about something like..
```
basemap ^1\.0\.[0-9]+$
```
and an if/else clause in the (admittedly already pretty cluttered) constructor of `VersionSpec`:
```python
elif spec.startswith("^") and spec.endswith("$"):
    self.spec = spec
    self.regex = re.compile(spec)
    self.match = self.regex_match_
```
That would enable people to use Regex in `pinned` file (detected by expicit start/end of string markers)..

### PR DESCRIPTION
**Pull request by megies**
Treat version specification strings that start with "^" and end with "$" as a regular expression. Fixes #2716.

The main point left to discuss is where to put the regex detection when interpreting a version specification, see https://github.com/conda/conda/commit/a39c400ce75be6570d644eb65c5dbdbd63554a43.

### CODE REVIEW

**Comment by kalefranz** [comment]
@megies Mind clicking that **clahub** Details link and filling that out?  It's new as of today, so you didn't miss anything.  Have a lot of work to go back and get agreement from previous contributors.  Have to start somewhere though.

**Comment by megies** [comment]
> @megies Mind clicking that clahub Details link and filling that out? It's new as of today, so you didn't miss anything. Have a lot of work to go back and get agreement from previous contributors. Have to start somewhere though.

Sure, no problem. Slick approach to have this tedious topic covered by a PR status check hook.

`You may provide support for free, for a fee, or not at all.`
Sounds a bit like "Any real number can be divided by 3 leaving a remainder, or not". ;-)

**Comment by megies** [comment]
> I'm concerned about preference here. Specifically, you put this before |, ,, and the pip inequalities. It seems to me it ought to be moved down below those three branches in preference.
> 
> Also, what if we change the and to an or? That way, people could do something like -g1234abcd$ to refer to a version with a git commit suffix.

mcg1969 replied 6 days ago
> 
> I should add: moving the regex test to after | and , is essential if you want to combine a regex match with a standard version match.

mcg1969 replied 6 days ago
> 
> For instance: suppose I want to accept any version after 1.2, or a specific patched version of 1.1. Then the match would be: >=1.2|^1.1.*-g1234abcd$

@mcg1969, regarding the main concern left on this, which is that putting the check for regex (startswith "^", endswith "$") at the top of the if/elif chain in the version spec `__new__()`, and thus putting it before the check for a pipe (`|`) (which is needed as a regex special character), potentially taking away the chance to do a logical `or`..

.. thinking about it again, I think it actually is no big deal. It would only affect a chain of `or`s that have a regex as first *and* last element of an `or` chain. That would mean you're looking at a real geek and (s)he should be able to figure out how to overcome this problem (e.g. by chaining another `or` with a non-regex fixed string condition that is never matched, like `^(some regex here)$|^(some other regex there)$|XXXXXXXXXX`)


In any case, the key point is to do proper documentation, if this gets the nod, I'll do a PR against http://conda.pydata.org/docs/spec.html?highlight=specification#package-match-specifications.

File: conda/version.py [hunk 1]
```diff
@@ -302,6 +302,10 @@ def __new__(cls, spec):
         self.spec = spec
         if isinstance(spec, tuple):
             self.match = self.all_match_ if spec[0] == 'all' else self.any_match_
+        elif spec.startswith("^") and spec.endswith("$"):
```
Comment (kalefranz) [inline conda/version.py]: How about something like this, which isn't quite python code exactly, and really sloppy even if it were


    elif re.match(r'(\^\S+?\$)', spec) as m:
        first = m.group()
        operator = spec[m.end()] if len(spec) > m.end() else None
        if operator is None:
            self.spec = first
            self.regex = re.compile(spec)
            self.match = self.regex_match_
        else:
            return VersionSpec((operator, tuple(first, spec[m.end()+1:])))

Comment (mcg1969) [inline conda/version.py]: What are you trying to achieve with this, exactly?

Comment (mcg1969) [inline conda/version.py]: @megies and I agreed that if someone is that determined to do full regex matching there's no need to handle the other operators. Just let it be standalone. After all, the other operators can be emulated in the regex itself.

Comment (kalefranz) [inline conda/version.py]: Make you both happy.  This would handle something like

    ^abc$|^def$,ghi|jkl,^mno$

Comment (kalefranz) [inline conda/version.py]: Moreover I really don't care either way. Just figured there was a way to give you both exactly what you were asking for. Then I read the rest of the conversation and realized you already agreed on something.

The PR is fine to me as is, or we can make it more robust as I suggested above. Completely indifferent.

Comment (megies) [inline conda/version.py]: Slick approach, and I kind of agree with the "why not if we can". I don't see any drawbacks right away either..

**Comment by kalefranz** [comment]
Superseded by #2933.  Pulled @megies commits and did just a bit of cleanup.

**Pull request closed without merging**